primitives/src/crh/mod.rs

```diff
@@ -23,20 +23,211 @@ pub trait FixedLengthCRH {
     fn evaluate(parameters: &Self::Parameters, input: &[u8]) -> Result<Self::Output, Error>;
 }
 
-pub trait FieldBasedHashParameters{
+/// Define parameters required to implement a hash function working with field arithmetics.
+/// TODO: Depending on the hash construction some parameters may be present and others not
+///       we should think about particularizing or generalizing this trait definition.
+pub trait FieldBasedHashParameters: Clone {
     type Fr: Field;
+
+    /// The rate of the hash function
+    const R: usize;
 }
 
 pub trait FieldBasedHash {
     type Data: Field;
     type Parameters: FieldBasedHashParameters<Fr = Self::Data>;
 
-    fn evaluate(input: &[Self::Data]) -> Result<Self::Data, Error>;
+    /// Initialize a Field Hash accepting inputs of constant length `input_size`:
+    /// any attempt to finalize the hash after having updated the Self instance
+    /// with a number of inputs not equal to `input_size` should result in an error.
+    /// Initialize the hash to a null state, or with `personalization` if specified.
+    fn init_constant_length(input_size: usize, personalization: Option<&[Self::Data]>) -> Self;
+
+    /// Initialize a Field Hash accepting inputs of variable length, adapting the padding
+    /// strategy according to the inputs being `mod_rate` or not: if the input is of
+    /// variable length but always multiple of the rate it is possible to avoid padding
+    /// without any security issue, saving constraints in SNARK applications;
+    /// If `mod_rate` is true, any attempt to finalize the hash after having updated the
+    /// Self instance with a number of inputs which are not multiple of the rate should
+    /// result in an error.
+    /// Initialize the hash to a null state, or with `personalization` if specified.
+    fn init_variable_length(mod_rate: bool, personalization: Option<&[Self::Data]>) -> Self;
+
+    /// Update the hash with `input`.
+    fn update(&mut self, input: Self::Data) -> &mut Self;
```

Comment:
Why do we only allow a single field element to be updated?
To me, this seems to be the wrong place to distinguish. Typically the update function of a Hash allows arbitrary length data.